compiler/fir/resolve/src/org/jetbrains/kotlin/fir/types/ConeTypeCompatibilityChecker.kt

```diff
@@ -0,0 +1,437 @@
+        val typeParameterOwner: FirClassLikeDeclaration<*>,
+        val mapping: Map<FirTypeParameterRef, BoundTypeArgument>
+    )
+
+    /** A single bound type argument to a type parameter declared in a class. */
+    private data class BoundTypeArgument(val type: ConeKotlinType, val variance: Variance)
+
+    /** Accumulated type arguments bound to a type parameter declared in a class. */
+    private data class BoundTypeArguments(
+        val upper: MutableSet<ConeClassLikeType>,
+        val lower: MutableSet<ConeClassLikeType>,
+        val compatibilityUpperBound: Compatibility
+    )
+
+    private fun ConeClassLikeType.toConcreteClass(ctx: ConeInferenceContext): ConcreteClass? {
+        return lookupTag.toConcreteClass(ctx)
+    }
+
+    private fun ConeClassLikeLookupTag.toConcreteClass(
+        ctx: ConeInferenceContext
+    ): ConcreteClass? = when (val klass = ctx.symbolProvider.getSymbolByLookupTag(this)?.fir) {
+        is FirTypeAlias -> klass.getAliasedClass(ctx)?.let { ConcreteClass.TypeAlias(klass, it, ctx) }
+        is FirClass<*> -> ConcreteClass.Class(klass, ctx)
+        else -> null
+    }
+
+    private sealed class ConcreteClass {
+        data class Class(val firClass: FirClass<*>, override val ctx: ConeInferenceContext) : ConcreteClass()
+        data class TypeAlias(val firTypeAlias: FirTypeAlias, val referencedClass: FirClass<*>, override val ctx: ConeInferenceContext) :
+            ConcreteClass()
+
+        abstract val ctx: ConeInferenceContext
+
+        val isInterface: Boolean
+            get() {
+                return when (this) {
+                    is Class -> firClass.isInterface
+                    is TypeAlias -> referencedClass.isInterface
+                }
+            }
+
+        val superConcreteClasses: Set<ConcreteClass> by lazy {
+            val firClass: FirClass<*> = when (this) {
+                is Class -> firClass
+                is TypeAlias -> referencedClass
+            }
+            firClass.superConeTypes.mapNotNull { it.lookupTag.toConcreteClass(ctx) }.toSet()
+        }
+
+        @OptIn(ExperimentalStdlibApi::class)
+        val thisAndAllSuperConcreteClasses: Set<ConcreteClass> by lazy {
+            val queue = ArrayDeque<ConcreteClass>()
+            queue.addLast(this)
+            buildSet {
+                add(this@ConcreteClass)
+                while (queue.isNotEmpty()) {
+                    val current = queue.removeFirst()
+                    val superTypes = current.superConcreteClasses
+                    queue.addAll(superTypes - this@buildSet)
+                    addAll(superTypes)
+                }
+            }
+        }
+
+        val isFinal: Boolean
+            get() {
+                return when (this) {
+                    is Class -> firClass.isFinal
+                    is TypeAlias -> referencedClass.isFinal
+                }
+            }
+
+        val hasPredefinedEqualityContract: Boolean
```

Comment:
Could you add a comment on what it means for these types to have a pre-defined equality contract?